### PR TITLE
Allow access to restricted system indices for reserved system roles

### PR DESCRIPTION
**Pull request by williamrandolph**
After we merged https://github.com/elastic/elasticsearch/pull/74212 , the Kibana team saw permissions errors at startup.

The cause of this is that all system indices are now restricted indices from the perspective of security, when previously the only restricted indices were security and async search indices. In that PR, we didn't update the reserved system roles so that the ones that access system indices have `allowRestrictedIndices = true`.

We can find the roles that need this change by adding system index patterns to the `TestRestrictedIndices` class and looking for failures in `ReservedRolesStoreTests`.

The fixes for the particular kibana issue are the updates to the `kibana_system` role.

### CODE REVIEW

**Comment by tylersmalley** [comment]
I verified this resolved the issue preventing the promotion of the nightly snapshot for Kibana.

File: x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/authz/store/ReservedRolesStore.java [hunk 1]
```diff
@@ -383,18 +383,20 @@ public static RoleDescriptor kibanaSystemRoleDescriptor(String name) {
...
                     .indices("apm-*")
-                    .privileges("read", "read_cross_cluster").build(),
+                    .privileges("read", "read_cross_cluster")
+                    .allowRestrictedIndices(true).build(),
                 // Data telemetry reads mappings, metadata and stats of indices
                 RoleDescriptor.IndicesPrivileges.builder()
                     .indices("*")
-                    .privileges("view_index_metadata", "monitor").build(),
+                    .privileges("view_index_metadata", "monitor")
+                    .allowRestrictedIndices(true).build(),
```
Comment (ywangd) [inline x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/authz/store/ReservedRolesStore.java:399]: This change allows the `kibana_system` user to have some access to the `.security` index. I don't think this is what we want.

**Comment by tvernum** [comment]
It looks to me like this issue slipped through because `TestRestrictedIndices` only contains security and async search indices.

Can we update it to include the complete set of out-of-the-box system indices?

File: x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/authz/store/ReservedRolesStore.java (hunk 1 above)
Comment (williamrandolph) [reply x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/authz/store/ReservedRolesStore.java]: I went through the list of system indices and pared down where I was allowing restricted indices to just the places where we are dealing with system indices, so I hope I've addressed this concern.

**Comment by williamrandolph** [comment]
@tvernum 

Thanks for the pointer. I think that is exactly the issue. I've manually added all the system index patterns to the `TestRestrictedIndices` class. I would really like to refactor this work so that we are referencing the actual string constants that the system index plugins use, but I think there's a good argument for pushing that to a follow-up and merging a fix for the Kibana team as soon as possible so that they can update their Elasticsearch snapshots.

File: x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/authz/store/ReservedRolesStore.java (hunk 1 above)
Comment (ywangd) [reply x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/authz/store/ReservedRolesStore.java]: Yes the change addressed my previous comment, thanks. But we still have another semantic change: before the #74212, this index privilege allow kibana to `monitor` and `view_index_metadata` for indices like `.watches`, `.triggered_watches`. These indices are now no longer accessible for `kibana_system`. Maybe this is the right thing to do. But I'd like to double check whether this is intentional.

**Review by ywangd — COMMENTED** [review]
The way how privileges are granted to restricted indices feels brittle to me. It requires checking and matching names from different places carefully (as well as adding them to tests). But maybe it is because this PR is a one-time mass update. It may feel alright for future smaller changes. Also I don't have a good suggestion for alternatives. 

I think we need a conclusion on `kibana_system`'s `monitor` and `view_index_metadata` access for `*`. Should we strictly maintain the previous semantic or are we ok to change? I am ok either way as long as we are conscious on the decision. Thanks!

**Review by ywangd — APPROVED** [review]
LGTM

> I think it could be argued that we should exclude other system indices from data telemetry

I tend to agree especially this change is 8.0 only. But I am not familiar with all system indices and how they should be surfaced by Kibana. I think it needs input from people who are more familiar with its usage (maybe a round of emails?). For this PR, I think maintaining the exact same semantic is the safest. 

Can I also suggest we add a few more assertions about `kibana_system` cannot view metadata or monitor `.security*` and `.async-search*`? The [current assertions](https://github.com/elastic/elasticsearch/blob/d36f24fbc367c3c51be81ba1c8481d0bd7b5a95e/x-pack/plugin/core/src/test/java/org/elasticsearch/xpack/core/security/authz/store/ReservedRolesStoreTests.java#L599-L600) only ensure no read or write access to the indices, but not metadata level access. I think this is also why the test didn't tripped on the first version of this PR.

**Pull request merged**